netcreateorg/netcreate-itest
# Infrastructure: Module Architecture Stage 3

This PR adds the beginning of our **testing framework** as well as some other quality of life features:

* NEW: The [vitest](https://vitest.dev/guide/) test suite is now integrated into the `_ur` build tool, making it possible to add **unit testing** to our addon libraries and run them all at once.
* NEW: The V2 **StateMgr** has been polished-up and ported from GEMSTEP. 
* FIX: While running the NetCreate, saving a file in `_ur` or `_ur_addons` will now force the application to be recompiled and reload. Before, you would have to manually stop/restart.
* ERROR REPORTING: Cleaner error output when a crash bug occurs in `NetCreate.jsx` or other root component. Before, it was difficult to tell what was happening.
* QOL: the integrated terminal will now properly `cd` to the project root directory, even if you are cloning a new terminal while inside the `_ur` directory. Before, it would open the current working directory and display an error.

### 0. INITIALIZING FOR TEST

* after pulling branch, open integrated terminal window
* `npm ci`

This will install the new `vitest` package.

### 1. TRY THE TESTING SYSTEM

* `cd _ur`
* `ur test`

You'll see some information about the current test. There is currently only on, located in `_ur/common/class-statemgr-test.ts`. The testing system will find _any_ `ts` file that ends with `*-test.ts` or starts with `test-*.ts`.

### 2. TEST NETCREATE and STATEMGR CLASS

* `cd ..`
* `npm run dev` and open localhost:3000
* In the console of NetCreate, look for this text indicating successful instantiation of the new StateMgr class
    ![image](https://github.com/netcreateorg/netcreate-itest/assets/2048828/186aa707-9158-41fc-90c0-2c871992068e)

### 3A. TEST AUTOMATIC REBUILD OF _UR CORE

* Make sure that the previous test is still running. We'll make a change to the StateMgr class in `_ur/common/class-statemgr.ts`
* Note in the current console what it looks like.
* Open `_ur/common/class-statemgr.ts` and find the `constructor()`
* Add `console.log('*** I WAS HERE')` and then save the file
* Quickly switch to the integrated VSCODE terminal and observe that the change has been detected, and the libraries are being recompiled without having to manually restart the server.
* Reload the webapp and check the console to see if `*** I WAS HERE` shows up.

### 3B. TEST AUTOMATIC REBUILD OF _UR ADDONS

* Make sure that the previous test is still running. We'll make a change to the `midi-core.ts`file `_ur_addons/midi/`
* See that the app recompiles as in test 3A

### 4. WITNESS ERROR HANDLING CHANGES

* In `NetCreate.jsx`, insert bad syntax in the middle of the constructor function (e.g. `BOO();`)
* On save, the NetCreate app will show an alert describing what to do, pausing the screen so you can open the console window before halting operation into the debugger. 
* Click the OK button.
* In the console window, switch back to the Javascript console output, and see if you can figure out where the error was found. 

### 5. TEST AUTOHOMING INTEGRATED TERMINAL

* Make sure there is only a single integrated terminal and `cd _ur`
* Click the "Split Terminal" button in the upper right of the terminal panel
* You should see that the new terminal is in your project root, with `vs_env` running without error

### OTHER INFORMATION

* split `ursys-lib` bundle into `ursys-core` and `ursys-addons`, which are loaded via `index.ejs`
* added `source-map` library for future help to deal with sourcemaps not pointing to useful locations for debugging. This is because the URSYS core and addon libraries are transformed twice, so the original source maps are not mapped correctly in NetCreate specifically. 
* added `class-op-sequencer` to `ursys-core` with test (unused)
